libgiza/typed_dict.py

```diff
@@ -0,0 +1,95 @@
+"""
+Provides the basis for a more strictly typed dictionary that ensures all keys
+and values will be of a specific type, and will throw exceptions if keys,
+values, or pairs of keys and values are of the wrong type or fail validation.
+"""
+
+import abc
+
+
+class TypedDict(dict):
+    """
+    An abstract base class definition that ensures that keys and values are of
+    the correct type. Requires users implement ``check_key()``,
+    ``check_value()`` and ``check_pair()`` methods that allow the these objects
+    to validate input.
+    """
+
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, key_type, value_type):
+        errors = []
+        if isinstance(key_type, type):
+            self.key_type = key_type
+        else:
+            errors.append("key_type ({0}) is not a type value".format(key_type))
+
+        if isinstance(value_type, type):
+            self.value_type = value_type
+        else:
+            errors.append("value_type ({0}) is not a type value".format(value_type))
+
+        if len(errors) > 0:
+            raise TypeError("; ".join(errors))
+
+    def __setitem__(self, key, value):
+        type_errors = []
+        value_errors = []
+
+        if isinstance(key, self.key_type):
+            value_errors.extend(self.check_key(key))
+        else:
+            try:
+                key = self.key_type(key)
```

Comment:
It's useful (i.e. for dicts and objects that can do the conversion it makes sense and and prevents needing to wrap custom default item setters all the time.) I expect that the constructor for most types will error in many cases. At the same time, numbers will convert implicitly to bools, and other types, which might be terrible too (but maybe not?)